prime-router/docs/proposals/0019-website-api-payload-structure.md

```diff
@@ -0,0 +1,332 @@
+| 400  | Bad Request           | Error; user submitted invalid/malformed data                                                     | Admin submits malformed Sender/Receiver settings data              
+| 401  | Unauthorized          | Error; user is not currently authorized to access the resource (may need to re-authenticate)     | Receiver requests deliveries with outdated Okta token              
+| 403  | Forbidden             | Error; user is never authorized to access the resource (re-authentication won't affect anything) | Receiver requests another Receiver's deliveries                    
+| 404  | Not Found             | Error; requested resource does not exist                                                         | Admin requests Receivers for a non-existent Organization           |
+| 500  | Internal Server Error | Unhandled exception                                                                              | Catch-all for all unhandled exceptions                             |
+
+NOTE: This table does not include common HTTP response codes commonly used as navigation signals, such as `301 Moved Permanently` or `308 Permanent Redirect`, which are not currently applicable to the website due to its single-page application (SPA) approach.  This may change in the future if we migrate to a more traditional website architecture.
+
+### HTTP methods
+
+There are no proposed changes to how we're using HTTP methods at the moment.  However, for all mutating methods (`POST`, `PUT`, `PATCH`, `DELETE`), we should return the entire modified record with all data through a `200 OK` rather than subsets of data or a `204 No Content`.  This is slightly atypical behavior, but it helps simplify request bookkeeping on the website.
+
+### Req/res structures
+
+Whether they include a single entity or a list of entities, **all responses should take the form of a JSON object as opposed to a JSON array**.  This is primarily for better scalability so that we can more easily add additional properties and metadata if we need to in the future.  (It incidentally has the added benefit of being more resistant to [JSON hijacking](https://haacked.com/archive/2008/11/20/anatomy-of-a-subtle-json-vulnerability.aspx/), but this is no longer a concern in modern browsers.)
+
+#### Resource req/res (`show` of an individual entity)
+
+Resource requests paths should take the form of `*/<resources>/<id>`, where:
+- `resources` is the entity view, facade, or database table
+- `id` is a publicly displayable identifier (should at least be an indexed field if querying from a table)
+
+Resource responses should take the form of a JSON object:
+
+```json
+{
+    "property1": ...,
+    "property2": ...,
+    "property3": ...,
+    // other resource data     
+    ...
+}
+```
+
+##### Example
+
+`GET /api/settings/organizations/ignore`
```

Comment:
> `resources` is the entity view, facade, or database table

This definition is not quite correct, right? If we look at the example of `GET /api/settings/organizations/ignore`, which you say is okay, "setting" is the table in the database and "organizations" is not anything, its just a logical concept. So we need to either clarify this definition (which I am in favor of) or we say this example is not correct (which I am not in favor of). I think what we should improve in this example is singularize "settings" and "organizations"

Reply:
Fair point!  This is based on the current [SettingsFunction.kt](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/src/main/kotlin/azure/SettingsFunctions.kt), which reads into an OrganizationAPI from `getOne` [here](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/src/main/kotlin/azure/SettingsFunctions.kt#L59).  BUT you raise a good point -- since settings are overloaded for different types (Organization, Sender, Receiver), I'm thinking it'll just be better to choose a different entity type altogether for clarity in this doc.

Regarding the singularization, are you proposing that we do something like `GET /api/setting/organization/ignore`?  If so, I disagree -- my feeling is that we should be aligning with standard REST practices regarding resources/collections so we can have a predictable pattern:

- `GET /resources`: fetch a list of resources
- `GET /resources/<resource_id>`: fetch a single resource
- `POST /resources`: create a new resource
- `(PUT|PATCH) /resources/<resource_id>`: update a single resource
- `DELETE/resources/<resource_id>`: delete a single resource

Reply:
re singularization: never mind, you are correct :)